### PR TITLE
K8SPSMDB-506: Fix restarting primary pod on smart update

### PR DESCRIPTION
**Pull request by egegunes**
[K8SPSMDB-506](https://jira.percona.com/browse/K8SPSMDB-506) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

We try to decide which pod is primary by running `rs.status()` and
getting `name` field of PRIMARY. If replset is not exposed `name` is
local FQDN of service and since service name and pod name is the same, we
can select the right pod. But if replset is exposed, node names are
external hostnames and selecting the primary pod is not straightforward.
We need to get the external address of each pod in the replset and
compare with primary host.

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| self-healing-chaos | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| operator-self-healing | passed |
| liveness | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| smart-update | passed |
| operator-self-healing-chaos | passed |
| pitr | passed |
| version-service | failed |
| users | passed |
| data-sharded | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/f308ee5d2f12b01cd42137bc88daed64acc40a71
image: `perconalab/percona-server-mongodb-operator:PR-719-f308ee5d`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| pitr-sharded | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/f308ee5d2f12b01cd42137bc88daed64acc40a71
image: `perconalab/percona-server-mongodb-operator:PR-719-f308ee5d`

**Review by qJkee — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| self-healing-chaos | passed |
| rs-shard-migration | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | failed |
| smart-update | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/102aabbda58280681cf39707ca583b0cc0108598
image: `perconalab/percona-server-mongodb-operator:PR-719-102aabbd`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| self-healing-chaos | passed |
| service-per-pod | failed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| operator-self-healing | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| operator-self-healing-chaos | passed |
| pitr | passed |
| pitr-sharded | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/c0f998cb4cca992d86dec3bcdd97832035b5cfc0
image: `perconalab/percona-server-mongodb-operator:PR-719-c0f998cb`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/c0f998cb4cca992d86dec3bcdd97832035b5cfc0
image: `perconalab/percona-server-mongodb-operator:PR-719-c0f998cb`

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| self-healing-chaos | failed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| operator-self-healing | passed |
| liveness | passed |
| upgrade | passed |
| operator-self-healing-chaos | failed |
| smart-update | passed |
| upgrade-sharded | passed |
| version-service | passed |
| pitr | passed |
| users | passed |
| data-sharded | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/736922c2109a361dd69f8cb5de661360bc6b49f9
image: `perconalab/percona-server-mongodb-operator:PR-719-736922c2`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | failed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| upgrade-consistency | passed |
| users | passed |
| demand-backup | passed |
| data-sharded | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| pitr-sharded | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/736922c2109a361dd69f8cb5de661360bc6b49f9
image: `perconalab/percona-server-mongodb-operator:PR-719-736922c2`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| self-healing-chaos | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| operator-self-healing | passed |
| liveness | passed |
| operator-self-healing-chaos | passed |
| smart-update | passed |
| version-service | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| users | passed |
| pitr | passed |
| data-sharded | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/0f023fdf5a5872cbb685841784c5a21c8326399d
image: `perconalab/percona-server-mongodb-operator:PR-719-0f023fdf`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| self-healing-chaos | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| version-service | passed |
| operator-self-healing | passed |
| upgrade-sharded | passed |
| operator-self-healing-chaos | passed |
| pitr | passed |
| users | passed |
| data-sharded | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/5d0a1d68d24666405dba9018c5cd00ab9edf5334
image: `perconalab/percona-server-mongodb-operator:PR-719-5d0a1d68`

**Review by delgod — APPROVED** [review]

File: pkg/controller/perconaservermongodb/smart.go [hunk 1]
```diff
@@ -3,11 +3,8 @@ package perconaservermongodb
 import (
 	"context"
 	"fmt"
-	"sort"
-	"strings"
```
Comment (qJkee) [inline pkg/controller/perconaservermongodb/smart.go:7]: strange work of goimports

File: pkg/psmdb/mongo/mongo.go [hunk 2]
```diff
@@ -71,7 +71,7 @@ func ReadConfig(ctx context.Context, client *mongo.Client) (RSConfig, error) {
 		return RSConfig{}, errors.Wrap(res.Err(), "replSetGetConfig")
 	}
 	if err := res.Decode(&resp); err != nil {
-		return RSConfig{}, errors.Wrap(err, "failed to decode to replSetGetConfig")
+		return RSConfig{}, errors.Wrap(err, "failed to decoge to replSetGetConfig")
```
Comment (qJkee) [inline pkg/psmdb/mongo/mongo.go:74]: decode is the right one =)

File: pkg/psmdb/mongo/mongo.go [hunk 3]
```diff
@@ -170,7 +170,7 @@ func WriteConfig(ctx context.Context, client *mongo.Client, cfg RSConfig) error
 	}
 
 	if err := res.Decode(&resp); err != nil {
-		return errors.Wrap(err, "failed to decode to replSetReconfigResponse")
+		return errors.Wrap(err, "failed to decoge to replSetReconfigResponse")
```
Comment (qJkee) [inline pkg/psmdb/mongo/mongo.go:173]: decode is the right one =)

Comment (egegunes) [reply pkg/psmdb/mongo/mongo.go]: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/de7cb57eba854c2446adfdc36261fbb71ebc1b04

File: pkg/psmdb/mongo/mongo.go (hunk 2 above)
Comment (egegunes) [reply pkg/psmdb/mongo/mongo.go]: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/de7cb57eba854c2446adfdc36261fbb71ebc1b04

File: pkg/controller/perconaservermongodb/smart.go (hunk 1 above)
Comment (egegunes) [reply pkg/controller/perconaservermongodb/smart.go]: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/de7cb57eba854c2446adfdc36261fbb71ebc1b04

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| self-healing-chaos | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| operator-self-healing | passed |
| smart-update | passed |
| operator-self-healing-chaos | passed |
| version-service | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| users | passed |
| pitr | passed |
| data-sharded | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/3bef4a162f70d4eeeaf3fb5729894f35ff3ef634
image: `perconalab/percona-server-mongodb-operator:PR-719-3bef4a16`

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| self-healing-chaos | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| operator-self-healing | passed |
| upgrade-sharded | passed |
| version-service | passed |
| operator-self-healing-chaos | passed |
| pitr | passed |
| users | passed |
| data-sharded | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/4797a04da8b1dd370f5981e5a5a0d7d2f0995326
image: `perconalab/percona-server-mongodb-operator:PR-719-4797a04d`

**Review by qJkee — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| self-healing | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| self-healing-chaos | failed |
| liveness | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | failed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | failed |
| smart-update | passed |
| version-service | passed |
| users | failed |
| pitr-sharded | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/668f36e02c6b2a5c54824cc48e32ba4017449f31
image: `perconalab/percona-server-mongodb-operator:PR-719-668f36e0`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| limits | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| data-sharded | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/719/commits/668f36e02c6b2a5c54824cc48e32ba4017449f31
image: `perconalab/percona-server-mongodb-operator:PR-719-668f36e0`

**Pull request merged**